### PR TITLE
Fix decoder omission highlighted by issue #89

### PR DESCRIPTION
**Pull request by IainCRobertson**
Would like to include in release 2.0.1 as this has not been officially published yet.

### CODE REVIEW

File: decoder.tex [hunk 1]
```diff
@@ -139,8 +140,8 @@ \section{Decoder pseudo code}
           #  final retired instruction
         inferred_address = TRUE
         return
-      if (te_inst.format == 3 and pc == address and
-        (branches == (is_branch(get_instr(pc)) ? 1 : 0)))
+      if (te_inst.format == 3 and pc == address and !unprocessed_branches(pc) and
+        (te_inst.privilege == privilege or is_return_from_trap(get_instr(last_pc))))
```
Comment (ved-rivos) [inline decoder.tex:144]: Q: the process_te_inst initialized privilege to te_inst.privilege. This code then tests te_inst.privilege to be equal to privilege. Will not not be always true?

File: decoder.tex [hunk 2]
```diff
@@ -258,14 +259,22 @@ \section{Decoder pseudo code}
...
   return FALSE
 
-# Determine if instruction is an uninferable discontinuity #
-function is_uninferable_discon (instr)
+# Determine if instruction is a return from trap #
+function is_return_from_trap (instr)
 
-  if (is_uninferable_jump(instr) or
-      (instr.opcode == URET)      or
+  if ((instr.opcode == URET)      or
```
Comment (ved-rivos) [inline decoder.tex:265]: may be a good opportunity to drop URET which was never ratified.

**Review by ved-rivos — COMMENTED** [review]
Noted a couple of questions in the PR.

File: decoder.tex (hunk 1 above)
Comment (IainCRobertson) [reply decoder.tex:144]: No.  'privilege' is the global variable whereas te_inst.privilege is the privilege value from the current te_inst message being processed.  This value gets copied to the global variable 4 lines from the end of the code on page 82 of the PDF, whereas the compare is happening in the follow_execution_path call a few lines before this.

File: decoder.tex (hunk 2 above)
Comment (IainCRobertson) [reply decoder.tex:265]: Yes, happy to take this out.

File: decoder.tex (hunk 1 above)
Comment (ved-rivos) [reply decoder.tex:144]: Thanks. I misread the placement of the `privilege` initialization. Second question - should the `privilege` initialization be conditional based on `te_inst` being a format 3 packet?

Comment (ved-rivos) [reply decoder.tex:144]: Last question about the initial state of the `privilege` global variable. It gets consumed before initialization on the first packet.

Comment (IainCRobertson) [reply decoder.tex:144]: No it doesn't.  follow_execution_path isn't called on the 1st packet because start_of_trace is TRUE.

Comment (ved-rivos) [reply decoder.tex:144]: looks good to me.

**Comment by IainCRobertson** [comment]
Thanks for the attention to detail Ved.

**Pull request merged**